pkg/generators/duck_type.go

```diff
@@ -0,0 +1,187 @@
+	}
+
+	return DefaultRequeueAfterSeconds
+}
+
+func (g *DuckTypeGenerator) GetTemplate(appSetGenerator *argoprojiov1alpha1.ApplicationSetGenerator) *argoprojiov1alpha1.ApplicationSetTemplate {
+	return &appSetGenerator.ClusterDecisionResource.Template
+}
+
+func (g *DuckTypeGenerator) GenerateParams(appSetGenerator *argoprojiov1alpha1.ApplicationSetGenerator, _ *argoprojiov1alpha1.ApplicationSet) ([]map[string]string, error) {
+
+	if appSetGenerator == nil {
+		return nil, EmptyAppSetGeneratorError
+	}
+
+	// Not likely to happen
+	if appSetGenerator.ClusterDecisionResource == nil {
+		return nil, nil
+	}
+
+	// ListCluster from Argo CD's util/db package will include the local cluster in the list of clusters
+	clustersFromArgoCD, err := utils.ListClusters(g.ctx, g.clientset, g.namespace)
+	if err != nil {
+		return nil, err
+	}
+
+	if clustersFromArgoCD == nil {
+		return nil, nil
+	}
+
+	// Read the configMapRef
+	cm, err := g.clientset.CoreV1().ConfigMaps(g.namespace).Get(g.ctx, appSetGenerator.ClusterDecisionResource.ConfigMapRef, metav1.GetOptions{})
+
+	if err != nil {
+		return nil, err
+	}
+
+	// Extract GVK data for the dynamic client to use
+	versionIdx := strings.Index(cm.Data["apiVersion"], "/")
+	kind := cm.Data["kind"]
+	resourceName := appSetGenerator.ClusterDecisionResource.Name
+
+	log.WithField("resourcename.kind.apiVersion", resourceName+"."+kind+"."+
+		cm.Data["apiVersion"]).Info("ResourceName.Kind.Group/Version Reference")
+
+	if kind == "" || resourceName == "" || versionIdx < 1 {
+		log.Warningf("kind=%v, resourceName=%v, versionIdx=%v", kind, resourceName, versionIdx)
+		return nil, errors.New("There is a problem with the apiVersion, kind or resourceName provided")
+	}
+
+	// Split up the apiVersion
+	group := cm.Data["apiVersion"][0:versionIdx]
+	version := cm.Data["apiVersion"][versionIdx+1:]
+	log.WithField("kind.group.version", kind+"."+group+"/"+version).Debug("decoded Ref")
+
+	duckGVR := schema.GroupVersionResource{Group: group, Version: version, Resource: kind}
+	duckResource, err := g.dynClient.Resource(duckGVR).Namespace(g.namespace).Get(g.ctx, resourceName, metav1.GetOptions{})
+
+	if err != nil {
+		log.WithField("GVK", duckGVR).Warningf("resource was not found with name=%v", resourceName)
+		return nil, err
+	}
+
+	if duckResource == nil || duckResource.Object["status"] == nil {
+		log.Warning("duck type status missing")
+		return nil, nil
+	}
+
+	log.WithField("duckResourceName", duckResource.GetName()).Debug("found duck status on resource")
+
+	// Override the duck type in the status of the resource
+	statusListKey := "clusters"
+
+	matchKey := cm.Data["matchKey"]
+
+	if cm.Data["statusListKey"] != "" {
+		statusListKey = cm.Data["statusListKey"]
+	}
+	if matchKey == "" {
+		log.WithField("matchKey", matchKey).Warning("matchKey not found in " + cm.Name)
+		return nil, nil
+
+	}
+
+	res := []map[string]string{}
+
+	if duckResource.Object["status"].(map[string]interface{})[statusListKey] != nil {
```

Comment:
Can probably make this a bit easier to grok with something like:
```go
clusterList := duckResource.Object["status"].(map[string]interface{})[statusListKey]
if clusterList != nil {
	for _, cluster := range clusterList.([]interface{}) {
// (etc)
```

Reply:
In keeping with the generate I used `clusterDecisions` 😉